File: SHM_16_Classes_with_smart_unique_and_XPointer.R
# After using raw pointers we switch now to smart pointers which take care of
# their lifetime administration and release memory as they go out of scope.
# We start with std::unique_ptr inside the C++ class.
# -----------------------------

Rcpp::sourceCpp("SHM_16_Classes_with_smart_unique_and_XPointer.cpp")

# create external pointer obj to our matrix
xptr_mat <- create_XPtr_for_R_obj(matrix(rnorm(12), 3, 4))
object.size(xptr_mat)

# check out the obj
print(xptr_mat)
xptr::xptr_address(xptr_mat)


# use external pointer as input to class obj creation where the
# the class uses the exact same data pointer to the obj
external_mat_admin <- new(External_Matrix_Administration, xptr_mat)
external_mat_admin$print_Matrix()
external_mat_admin$add_42_to_Matrix()
external_mat_admin$print_Matrix()
external_mat_admin$return_Matrix()
external_mat_admin$add_number_to_Matrix(42.42)
external_mat_admin$print_Matrix()
external_mat_admin$return_Matrix()
external_mat_admin$add_number_to_Matrix(42L)
external_mat_admin$print_Matrix()
external_mat_admin$return_Matrix()

# object no longer needed: delete it, release pointers and memory
rm(external_mat_admin)
gc()

# check out the obj
xptr::xptr_address(xptr_mat)
print(xptr_mat)

# release memory
xptr::xptr_clear(xptr_mat)
xptr::xptr_address(xptr_mat)
print(xptr_mat)
xptr::is_null_xptr(xptr_mat)

# the memory is now released, delete pointer obj
rm(xptr_mat)
gc()

File: SHM_16_Classes_with_smart_unique_and_XPointer.cpp
// [[Rcpp::depends(RcppArmadillo)]]
#include <iostream>
#include <memory>
#include <RcppArmadillo.h>


class External_Matrix_Administration {
  
public:
  
  // ctor
  // --------------------
  External_Matrix_Administration (Rcpp::XPtr<arma::mat> xptr_A) {
    Rcpp::Rcout << "XPtr from R: " << xptr_A << "\n";
    Rcpp::Rcout << *xptr_A << "\n";
    this->armaMat = std::unique_ptr<arma::mat>(xptr_A);
    Rcpp::Rcout << "Smart pointer in C++ taking the address: " << armaMat.get() << "\n";
    Rcpp::Rcout << *armaMat << "\n";
  };
  
  // dtor
  // --------------------
  ~External_Matrix_Administration() = default;

  // Member method
  // --------------------
  void print_Matrix() const{
    armaMat->print();
    Rcpp::Rcout << std::endl;
  }
  
  void add_42_to_Matrix() {
    *armaMat += 42;
  }
  
  void add_number_to_Matrix(const double number) {
    *armaMat += number;
  }

  arma::mat return_Matrix() const{
    return *armaMat;
  }
  
private:
  // Attribute
  // --------------------
  std::unique_ptr<arma::mat> armaMat{nullptr};

};


Rcpp::XPtr<arma::mat> create_XPtr_for_R_obj(arma::mat& A) {
  arma::mat* armaMat = new arma::mat(A.memptr(), A.n_rows, A.n_cols, true, false);
  return Rcpp::XPtr<arma::mat> (armaMat);
}


RCPP_MODULE(mod_External_Matrix_Admin) {
  using namespace Rcpp;

  function("create_XPtr_for_R_obj",
           &create_XPtr_for_R_obj,
           List::create(_["mat"]), "Create external pointer of R matrix obj");
  

  Rcpp::class_<External_Matrix_Administration>("External_Matrix_Administration")
   .constructor<Rcpp::XPtr<arma::mat>>("Creates new class obj.")
   .method("print_Matrix", &External_Matrix_Administration::print_Matrix, "Prints matrix.")
   .method("add_42_to_Matrix", &External_Matrix_Administration::add_42_to_Matrix, "Add 42 to matrix.")
   .method("return_Matrix", &External_Matrix_Administration::return_Matrix, "Return matrix to R.")
   .method("add_number_to_Matrix", &External_Matrix_Administration::add_number_to_Matrix, "Add number to Matrix.")
  ;
  
}